rails 5 + api v2 + raml api docs

devvmh opened the pull request:
Update to rails 5 and active_model_serializers 0.10.x. Major changes are:

- Gemfile is updated, and ruby-version changes to 2.3.0
- changes for sprockets 4: add a manifest file and fix a module.exports javascript error
- copy the config directory from a blank rails 5 app and merge the changes into our directory
- fix one ActiveRecord error in a controller (I think due to rails 5 changes)
- Remove `embed :ids, include: true` from serializers, because it's not supported in the new version of active model serializers
- Migrate to api/v2 because of breaking changes to the format
- RAML api documentation & tests to ensure our documentation matches reality (!!)

### Motivation for rails 5 upgrade

1. ActionCable, for better web sockets integration
2. NOT sprockets 4, though it's a nice bonus

We are using webpack now directly, so I think we don't need to use babel-transpiler (a sprockets 4 package which seems poorly maintained) or browserify-rails (a great package which makes dev a lot slower, annoyingly enough).

#### TODO:

- [x] Make Travis great again
  - [ ] small problem with `app/models/user_preference.rb` has been resolved with a workaround. This doesn't block release but it would be nice to better understand the problem.
- [x] Create dummy endpoints for api/v1 with a deprecation notice and a link to the new docs
- [x] Create api/v2
  - [x] topics
  - [x] synapses
  - [x] maps
  - [x] mappings
  - [x] tokens
- [x] RAML documentation for api/v2 (see https://www.devinhoward.ca/share/apidocs.html)
  - [x] topics
  - [x] synapses
  - [x] maps
  - [x] mappings
  - [x] tokens
- [x] rspec tests for api/v2 endpoints, (real responses match schema, schema matches examples in docs)
  - [x] topics
  - [x] synapses
  - [x] maps
  - [x] mappings
  - [x] tokens
- [ ] set up raml2html for 1.0
- [ ] upgrade raml syntax to 1.0
- [ ] QA testing on metamaps.devinhoward.ca

Connoropolous commented:
@devvmh taking a look through and seeing all the API documentation! so great. I'm guessing there's a way to auto-generate the documentation... does that come out of the box with RAML? Can we make it a section of the website?

devvmh commented:
Yes, if we use RAML 0.8 we can use `npm install raml2html` to generate some slick documentation. I'd like to use 1.0 though (only for future proofing, not for any particular features) so I need to find a tool that works with 1.0.

The existing raml I have is compatible with 0.8 though so I can change the version ID in the api.raml and get a cool file, I'll try to upload it somewhere

devvmh commented on doc/api/api.raml:
```diff
@@ -0,0 +1,31 @@
+      settings:
+        authorizationUri: https://metamaps.cc/api/v2/oauth/authorize
+        accessTokenUri: https://metamaps.cc/api/v2/oauth/token
+        authorizationGrants: [ authorization_code, password, client_credentials, implicit, refresh_token ]
```
@connor without diving too deep into raml, does this look right for out implementation? are we using OAuth2 or OAuth1? do you know what values authorizationGrants should have?

I basically just copy pasted this from http://raml.org/developers/whats-new-raml-10

Connoropolous replied:
OAuth2 yeah :)

devvmh replied:
what about this part

    authorizationGrants: [ authorization_code, password, client_credentials, implicit, refresh_token ]

Do you know what it means/what I should write on this line?

Connoropolous replied:
sorry about the delay. for this, we just want authorization_code and client_credentials. 

these are the defaults that doorkeeper suggests, so I went with those, as per: 
https://github.com/metamaps/metamaps/blob/develop/config/initializers/doorkeeper.rb#L83

devvmh commented:
Though there is more work to do, this could be merged any time. Connor, once you're back, maybe we can do a high intensity testing day on Heroku

devvmh commented on app/controllers/api/v2/restful_controller.rb:
```diff
@@ -0,0 +1,179 @@
+          end
+        end
+        return builder
+      end
```
Should we send serachable_columns, search_by_q, and order_by_sort back to snorlax as a PR? maybe.

devvmh commented on app/controllers/api/v2/sessions_controller.rb:
```diff
@@ -0,0 +1,20 @@
+      end
+    end
+  end
+end
```
I should remove this file or document the API, or decide consciously to do neither.

devvmh commented on app/serializers/api/v2/event_serializer.rb:
```diff
@@ -0,0 +1,18 @@
+      end
+    end
+  end
+end
```
should this have an API endpoint?

Connoropolous replied:
eventually, yeah :) it is ok to add API endpoints without changing the version yeah? because it's not breaking?

devvmh replied:
Definitely. 

In node, you have versioning like x.y.z. If you remove/change an existing API, you bump the major version (x). If you add a new API, you bump the minor version (y). If you just change internals/bugfixes, you only need to bump the patch version (z).

I don't propose we do minor/patch versions, but I do think we should follow the no removals/changes to existing APIs once this is released.

devvmh replied:
Which means of course, that decisions like

- data envelopes
- which columns are searched by ?q=
- removing :updated_at from the token_serializer

Are locking us in until we decide we need a v3

devvmh commented on config/initializers/cookies_serializer.rb:
```diff
@@ -0,0 +1,5 @@
+
+# Specify a serializer for the signed and encrypted cookie jars.
+# Valid options are :json, :marshal, and :hybrid.
+Rails.application.config.action_dispatch.cookies_serializer = :hybrid
```
Rails 3 used "marshal" encoding for cookies on the users' computers. We missed upgrading to json in rails 4, but can upgrade them now.

However, if we end up reverting this branch, we will have to make sure we leave the serializer as :json or :hybrid, or users will get server errors when they send marshal-encoded cookies

Once a user with a cookie visits our server one time, the hybrid encoder will change their cookie to be JSON instead, so at some point we can move to being all json all the time.

Connoropolous replied:
noted.

devvmh commented:
API documentation is generated, by faking 0.8 format for now, here: https://www.devinhoward.ca/share/apidocs.html

Obviously this page would need a lot of work to be really good, but it is our starting point. In particular, the schemas should be de-emphasized and the examples should be put at the forefront. We'll also need tutorials for a few tasks (e.g. accessing the API using Chrome/Firefox/Edge/Safari 10 & the fetch api)

Connoropolous commented:
@poietic check this out! https://www.devinhoward.ca/share/apidocs.html

harlantwood commented on app/models/user_preference.rb:
```diff
@@ -3,8 +3,12 @@ class UserPreference
+        array.push(m.id.to_s)
+      end
+    rescue ActiveRecord::StatementInvalid
+      Rails.logger.warn("TODO: remove this travis workaround in user_preference.rb")
```
In what sense is this a Travis workaround?  Curious.

devvmh replied:
I cant recall precisely when the error happens, so it may be just an "cant
initialize an empty database" workaround, but when you try to load the
schema it fails because this class tries to initialize and it cant find the
Metacode table.

It definitely happens on every travis run, and it might also happen for
people initializing a local db.

On Tue, Sep 13, 2016, 11:22 AM Harlan T Wood <notifications@github.com>
wrote:

> In app/models/user_preference.rb
> <https://github.com/metamaps/metamaps/pull/593#discussion_r78491248>:
>
> > @@ -3,8 +3,12 @@ class UserPreference
> >
> >    def initialize
> >      array = []
> > -    Metacode.all.find_each do |m|
> > -      array.push(m.id.to_s)
> > +    begin
> > +      Metacode.all.find_each do |m|
> > +        array.push(m.id.to_s)
> > +      end
> > +    rescue ActiveRecord::StatementInvalid
> > +      Rails.logger.warn("TODO: remove this travis workaround in user_preference.rb")
>
> In what sense is this a Travis workaround? Curious.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/metamaps/metamaps/pull/593/files/f4ac81b0b71e676d59fb6f3f2c8a14d9dea0f62b..1c38865db3d823e7da9686d701aaff8265236e20#r78491248>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABVELNjgm-B_BKj07AGQqYRz836th3i3ks5qphb0gaJpZM4JeexP>
> .
>

harlantwood replied:
I think if you do

bundle exec rake db:setup

Instead of

bundle exec rake db:schema:load

In Travis, it will load the db/seeds and fix the issue.

devvmh replied:
No, i tried that. The issue is that it autoloads the UserPreference constant, and for some reason calls initialize before the `metacodes` table is even created in Postgres. So both of the commands you list above will fail without this workaround.

devvmh commented:
We could maybe bump the version to 3.0 when we do this, given that we

1. updated to rails 5
2. updated API
3. recently added shared maps
4. recently overhauled the UI a bit

Connoropolous commented:
hitting /maps/new gives me the 'we're sorry, but something went wrong' page

devvmh commented:
Ah that's fixed in the "fix starred map error" commit; I'll redeploy so it gets that fix online.

poietic commented:
wow I'm really loving this!! Amazing! this feels so much like...a
platform's API docs guys :) hehehehe

On Tue, Sep 6, 2016 at 9:00 AM, Connor Turland <notifications@github.com>
wrote:

> @poietic <https://github.com/poietic> check this out!
> https://www.devinhoward.ca/share/apidocs.html
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/metamaps/metamaps/pull/593#issuecomment-244999440>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABXJ7knj8oPBMz9mITllczsEPPxRKoNrks5qnY4kgaJpZM4JeexP>
> .
>



-- 
Ishan M. Shapiro
Creative Entrepreneur
Metamaps.cc <http://metamaps.cc/> | Notthisbody <http://notthisbody.com> |
Futurescouts.co
@metamapps <http://twitter.com/metamapps> | @notthisbody
<http://twitter.com/notthisbody> | @futurescouts
<http://twitter.com/futurescouts>
<http://cargo.notthisbody.com>
